### ISSUE TITLE
Find navigation data (file, line no.) for tests written in .razor files

### ISSUE THREAD

**Issue by egil** [microsoft/vstest#2668]
## Description - feature request

**TLDR:** Read `#line` pragmas in generated files when discovering test method positions, to make it possible to use the "navigate to test by double clicking it in Test Explorer" for tests written in files that transpilled into .cs files, e.g. .razor files.

**Background:** With [bUnit, a test library for Blazor components](https://github.com/egil/bunit), it is possible to write regular unit tests in .razor files, e.g. `MyTest.razor`:

```razor
@using Xunit
@code
{
    [Fact]
    public void Test() { ... }

    [Theory]
    [InlineData("foo")]
    public void DataDrivenTest(string input) { ... }
}
```

These .razor files are compiled by the Blazor compiler into .cs and placed in the `obj\Debug\net5.0\Razor` folder, from where the tests are discovered and runnable from the Test Explorer in Visual Studio.

**However, the "navigate to test by double clicking it in Test Explorer"-feature doesn't work**, because the tests are discovered to be in e.g. `obj\Debug\net5.0\Razor\MyTest.razor.g.cs`.

The generated .cs file for the example above look roughly like this:

```C#
#pragma checksum "C:\source\bUnit\examples\MyTest.razor" "{ff1816ec-aa5e-4d10-87f7-6f4963833460}" "962b7f7bc4a3b12b6352aa215e782e2ff4708a2b"
// <auto-generated/>
#pragma warning disable 1591
namespace Bunit.RazorTesting
{
    // .. removed imports for brevity.
#line default
#line hidden
#nullable disable
    public partial class MyTest : TestContext
    {
        #pragma warning disable 1998
        protected override void BuildRenderTree(Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder __builder)
        {
        }
        #pragma warning restore 1998
#nullable restore
#line 3 "C:\source\bUnit\examples\MyTest.razor"
 
	[Fact]
	public void Test() { /* ... */ }

	[Theory]
	[InlineData("foo")]
	public void DataDrivenTest(string input) { /* ... */ }

#line default
#line hidden
#nullable disable
    }
}
#pragma warning restore 1591
```

As we can see, the `#line` pragma is added to it, so my suggestion is that the [VSTest navigation classes](https://github.com/microsoft/vstest/tree/master/src/Microsoft.TestPlatform.ObjectModel/Navigation) are tweaked to use those to find the "real" test method, when they are available.

That would make this "just work" for xUnit tests, [according to @bradwilson](https://github.com/xunit/xunit/discussions/2199). I am not sure if NUnit or MSTest is similarly using the `DiaSession` type as xUnit, but if they are, it would also just work in their cases.

Update: it looks as if [NUnit is also using DiaSession](https://github.com/nunit/nunit3-vs-adapter/blob/master/src/NUnitTestAdapter/NavigationDataProvider.cs) in their code.

## Environment

- Windows 10, Visual Studio 2019, .NET Core/Blazor/.NET5/.NETCoreApp3.1/bUnit/xUnit

**Comment by egil** [comment]
It seems this actually work in some cases as expected, so this might be a non-issue, but I will investigate further and report back.

**Comment by nohwnd** [comment]
Please do report back if you still think this is an issue.

**Comment by egil** [comment]
Ah yes, thanks, sorry for forgetting about this. It does indeed seem to work with the latest versions of VS.
